esphome/components/tsl2591/tsl2591.cpp

```diff
@@ -0,0 +1,295 @@
+
+namespace esphome {
+namespace tsl2591 {
+
+static const char *const TAG = "tsl2591.sensor";
+
+// Various constants used in TSL2591 register manipulation
+#define TSL2591_COMMAND_BIT (0xA0)      // 1010 0000: bits 7 and 5 for 'command normal'
+#define TSL2591_ENABLE_POWERON (0x01)   // Flag for ENABLE register, to enable
+#define TSL2591_ENABLE_POWEROFF (0x00)  // Flag for ENABLE register, to disable
+#define TSL2591_ENABLE_AEN (0x02)       // Flag for ENABLE register, to turn on ADCs
+
+// TSL2591 registers from the datasheet. We only define what we use.
+#define TSL2591_REGISTER_ENABLE (0x00)
+#define TSL2591_REGISTER_CONTROL (0x01)
+#define TSL2591_REGISTER_DEVICE_ID (0x12)
+#define TSL2591_REGISTER_STATUS (0x13)
+#define TSL2591_REGISTER_CHAN0_LOW (0x14)
+#define TSL2591_REGISTER_CHAN0_HIGH (0x15)
+#define TSL2591_REGISTER_CHAN1_LOW (0x16)
+#define TSL2591_REGISTER_CHAN1_HIGH (0x17)
+
+void TSL2591Component::enable() {
+  // Enable the device by setting the control bit to 0x01. Also turn on ADCs.
+  this->write_byte(TSL2591_COMMAND_BIT | TSL2591_REGISTER_ENABLE, TSL2591_ENABLE_POWERON | TSL2591_ENABLE_AEN);
+}
+
+void TSL2591Component::disable() {
+  if (this->power_save_mode_enabled_) {
+    // Disable the device by setting the control bit to 0x00
+    this->write_byte(TSL2591_COMMAND_BIT | TSL2591_REGISTER_ENABLE, TSL2591_ENABLE_POWEROFF);
+  }
+}
+
+void TSL2591Component::disable_internal_() {
+  if (this->power_save_mode_enabled_) {
+    this->disable();
+  }
+}
+
+void TSL2591Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up TSL2591...");
+  auto id = this->read_byte(TSL2591_COMMAND_BIT | TSL2591_REGISTER_DEVICE_ID);
+  if (id != 0x50) {
+    ESP_LOGD("ERROR", "Could not find a TSL2591 Sensor. Did you configure I2C with the correct address?");
+    this->mark_failed();
+  } else {
+    this->disable_internal_();
+  }
+}
+
+void TSL2591Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "TSL2591:");
+  LOG_I2C_DEVICE(this);
+
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with TSL2591 failed earlier, during setup");
+    return;
+  }
+
+  ESP_LOGCONFIG(TAG, "  Name: %s", this->name_);
+  TSL2591Gain raw_gain = this->gain_;
+  int gain = 0;
+  std::string gain_word = "unknown";
+  switch (raw_gain) {
+    case TSL2591_GAIN_LOW:
+      gain = 1;
+      gain_word = "low";
+      break;
+    case TSL2591_GAIN_MED:
+      gain = 25;
+      gain_word = "medium";
+      break;
+    case TSL2591_GAIN_HIGH:
+      gain = 400;
+      gain_word = "high";
+      break;
+    case TSL2591_GAIN_MAX:
+      gain = 9500;
+      gain_word = "maximum";
+      break;
+  }
+  ESP_LOGCONFIG(TAG, "  Gain: %dx (%s)", gain, gain_word.c_str());
+  TSL2591IntegrationTime raw_timing = this->integration_time_;
+  int timing_ms = (1 + raw_timing) * 100;
+  ESP_LOGCONFIG(TAG, "  Integration Time: %d ms", timing_ms);
+  ESP_LOGCONFIG(TAG, "  Power save mode enabled: %s", this->power_save_mode_enabled_ ? "True" : "False");
```

Comment:
```suggestion
  ESP_LOGCONFIG(TAG, "  Power save mode enabled: %s", ONOFF(this->power_save_mode_enabled_));
```
For consistency with what other components use :)